### PR TITLE
fix java build failure

### PR DESCRIPTION
**Pull request by zhijunfu**
## Why are these changes needed?

Recently java build has been broken by the removal of ActorHandle (#5889), then there are a couple of core worker changes afterwards that removes `TaskInterface`, `TaskExecution`, `ObjectInterface` from `CoreWorker` which didn't update Java worker. 

This PR updates Java side accordingly to fix the build issue. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -719,4 +719,29 @@ Status CoreWorker::BuildArgsForExecutor(const TaskSpecification &task,
   return status;
 }
 
+Status CoreWorker::ActorLanguage(const ActorID &actor_id, Language *language) const {
```
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.cc]: These methods should not be exposed on the `CoreWorker`. You can instead call `CoreWorker::GetActorHandle`, then call the relevant method on the returned actor handle.

**Review by stephanie-wang — APPROVED** [review]
Thanks!

**Review by stephanie-wang — CHANGES_REQUESTED** [review]
Sorry, meant to request changes on the last comment.

File: src/ray/core_worker/core_worker.cc (hunk 1 above)
Comment (zhijunfu) [reply src/ray/core_worker/core_worker.cc]: OK, updated.

File: src/ray/core_worker/core_worker.cc [hunk 2]
```diff
@@ -229,13 +229,22 @@ void CoreWorker::RunIOService() {
 void CoreWorker::SetCurrentTaskId(const TaskID &task_id) {
   worker_context_.SetCurrentTaskId(task_id);
   main_thread_task_id_ = task_id;
+  // NOTE(zhijunfu): AsyncUnsubscribe() is async and cannot guarantee
+  // that no actor update notifications will be received after it's called.
+  // If the worker receives a task which again adds actor handles for the
+  // same actors, then it's possible that there are two callbacks invoked
+  // for the same actor ID, which would crash when the second one tries to
+  // append the rpc client for the actor. So disable the logic below until
```
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.cc]: Thanks! But why not just keep the original rpc client for now?

File: src/ray/core_worker/core_worker.h [hunk 3]
```diff
@@ -321,6 +321,13 @@ class CoreWorker {
                                const std::vector<std::shared_ptr<Buffer>> &metadatas,
                                std::vector<std::shared_ptr<RayObject>> *return_objects);
 
+  /// Get a handle to an actor.
+  ///
+  /// \param[in] actor_id The actor handle to get.
+  /// \param[out] actor_handle A handle to the requested actor.
+  /// \return Status::Invalid if we don't have this actor handle.
+  Status GetActorHandle(const ActorID &actor_id, ActorHandle **actor_handle) const;
```
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.h]: I'm fine with it, but just curious, why does this need to be public for java but not python?

File: src/ray/core_worker/core_worker.cc (hunk 2 above)
Comment (zhijunfu) [reply src/ray/core_worker/core_worker.cc]: Could you please explain more on your suggestion. The issue is that `ConnectAndSendPendingTasks` is called twice and at the second time it crashes before there's already a rpc client for this `actor_id` thus the emplace fails.

File: src/ray/core_worker/core_worker.h (hunk 3 above)
Comment (zhijunfu) [reply src/ray/core_worker/core_worker.h]: I think it's because certain information is only required for java: e.g. `ActorLanguage`, `FunctionDescriptor`

File: src/ray/core_worker/core_worker.cc (hunk 2 above)
Comment (stephanie-wang) [reply src/ray/core_worker/core_worker.cc]: Right, I was saying that we could just keep the original rpc client instead of trying to do the emplace.

Anyway, I think this is fixed now that we merged #6191. You can see the code [here](https://github.com/ray-project/ray/blob/master/src/ray/core_worker/transport/direct_actor_transport.cc#L120). Can you uncomment this and see if it works now?

File: src/ray/core_worker/core_worker.cc [hunk 4]
```diff
@@ -524,7 +533,7 @@ Status CoreWorker::SubmitTask(const RayFunction &function,
   BuildCommonTaskSpec(
       builder, worker_context_.GetCurrentJobID(), task_id,
       worker_context_.GetCurrentTaskID(), next_task_index, GetCallerId(), function, args,
-      task_options.num_returns, task_options.resources, {},
+      task_options.num_returns, task_options.resources, std::unordered_map<std::string, double>{},
```
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.cc]: Is this needed?

File: src/ray/core_worker/core_worker.cc [hunk 5]
```diff
@@ -586,9 +595,10 @@ Status CoreWorker::SubmitActorTask(const ActorID &actor_id, const RayFunction &f
   const TaskID actor_task_id = TaskID::ForActorTask(
       worker_context_.GetCurrentJobID(), worker_context_.GetCurrentTaskID(),
       next_task_index, actor_handle->GetActorID());
+  std::unordered_map<std::string, double> empty_resources;
   BuildCommonTaskSpec(builder, actor_handle->CreationJobID(), actor_task_id,
                       worker_context_.GetCurrentTaskID(), next_task_index, GetCallerId(),
-                      function, args, num_returns, task_options.resources, {},
+                      function, args, num_returns, task_options.resources, empty_resources,
```
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.cc]: Is this needed?

File: src/ray/core_worker/core_worker.cc (hunk 2 above)
Comment (kfstorm) [reply src/ray/core_worker/core_worker.cc]: Reverted.

File: src/ray/core_worker/core_worker.cc (hunk 4 above)
Comment (kfstorm) [reply src/ray/core_worker/core_worker.cc]: Reverted.

File: src/ray/core_worker/core_worker.cc (hunk 5 above)
Comment (kfstorm) [reply src/ray/core_worker/core_worker.cc]: Reverted.

**Comment by kfstorm** [comment]
Due to https://github.com/ray-project/ray/issues/6329, the test case `testPassActorAsParameter` will hang in direct actor call mode. Disable it for now. Will enable it once the issue is fixed.

**Comment by kfstorm** [comment]
@stephanie-wang Can you revisit this PR?

File: src/ray/common/task/task_spec.h [hunk 6]
```diff
@@ -180,6 +180,7 @@ class TaskSpecification : public MessageWrapper<rpc::TaskSpec> {
   /// Cached scheduling class of this task.
   SchedulingClass sched_cls_id_;
 
+  static std::mutex sched_mutex_;
```
Comment (stephanie-wang) [inline src/ray/common/task/task_spec.h]: Why is this necessary?

File: src/ray/core_worker/core_worker.cc [hunk 7]
```diff
@@ -716,13 +717,16 @@ Status CoreWorker::SerializeActorHandle(const ActorID &actor_id,
 }
 
 bool CoreWorker::AddActorHandle(std::unique_ptr<ActorHandle> actor_handle) {
+  std::lock_guard<std::mutex> lock(actor_handles_mutex_);
```
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.cc]: Why was this not necessary before?

File: src/ray/core_worker/core_worker.h [hunk 8]
```diff
@@ -557,6 +556,8 @@ class CoreWorker {
   // Interface to submit non-actor tasks directly to leased workers.
   std::unique_ptr<CoreWorkerDirectTaskSubmitter> direct_task_submitter_;
 
+  mutable std::mutex actor_handles_mutex_;
```
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.h]: If you're going to add a lock, please use absl::Mutex and use thread annotations to show which fields are guarded by the mutex.

File: src/ray/common/task/task_spec.h (hunk 6 above)
Comment (kfstorm) [reply src/ray/common/task/task_spec.h]: Because in Java worker it's possible to sumbit tasks concurrently. So we need to protect these static fields of `TaskSpecification` which could be mutated in the constructor of `TaskSpecification`.

File: src/ray/core_worker/core_worker.cc (hunk 7 above)
Comment (kfstorm) [reply src/ray/core_worker/core_worker.cc]: Previously we don't store `ActorHandle` instances in a map, the instances are hold in user code. Due to multi-threading, now we need to protect the actor handles map.

File: src/ray/core_worker/core_worker.h (hunk 8 above)
Comment (kfstorm) [reply src/ray/core_worker/core_worker.h]: Sure, I can do that.

Comment (kfstorm) [reply src/ray/core_worker/core_worker.h]: Done.

File: java/runtime/src/main/java/org/ray/runtime/task/ArgumentsBuilder.java [hunk 9]
```diff
@@ -22,10 +23,13 @@
    */
   private static final int LARGEST_SIZE_PASS_BY_VALUE = 100 * 1024;
 
+  private static final NativeRayObject PYTHON_DUMMY_TYPE = ObjectSerializer
+      .serialize("__RAY_DUMMY__".getBytes());
```
Comment (raulchen) [inline java/runtime/src/main/java/org/ray/runtime/task/ArgumentsBuilder.java]: Can you add a comment here and in `signature.py`, explaining that both need to be sync'ed?

File: java/test/src/main/java/org/ray/api/test/ActorTest.java [hunk 10]
```diff
@@ -96,6 +94,7 @@ static int testActorAsFieldOfParameter(List<RayActor<Counter>> actor, int delta)
     return res.get();
   }
 
+  @Test(enabled = false)
```
Comment (raulchen) [inline java/test/src/main/java/org/ray/api/test/ActorTest.java]: add a TODO here?

File: java/runtime/src/main/java/org/ray/runtime/task/ArgumentsBuilder.java (hunk 9 above)
Comment (raulchen) [inline java/runtime/src/main/java/org/ray/runtime/task/ArgumentsBuilder.java]: Can you add a comment here and in `signature.py`, explaining that both need to be sync'ed?

File: src/ray/common/task/task_spec.h (hunk 6 above)
Comment (raulchen) [reply src/ray/common/task/task_spec.h]: Theoretically, Python should also have this issue. If we haven't run into it, it's probably because we are not releasing GIL properly.

File: java/runtime/src/main/java/org/ray/runtime/actor/NativeRayActor.java [hunk 11]
```diff
@@ -33,75 +44,64 @@ public NativeRayActor(long nativeActorHandle) {
...
+    Preconditions.checkState(runtime instanceof RayNativeRuntime);
+    nativeCoreWorkerPointer = ((RayNativeRuntime)runtime).getNativeCoreWorkerPointer();
+
+    actorId = nativeDeserialize(nativeCoreWorkerPointer, (byte[]) in.readObject());
   }
 
   @Override
   protected void finalize() {
-    nativeFree(nativeActorHandle);
+    // TODO(zhijunfu): do we need to free the ActorHandle in core worker?
```
Comment (raulchen) [inline java/runtime/src/main/java/org/ray/runtime/actor/NativeRayActor.java]: @stephanie-wang It seems that we don't do this in Python either? If I understand it correctly, we only clean up all handles at the end of each non-actor task. 
I think this would lead to memory leak in some cases. E.g., an actor with a method that takes an actor handle as parameter.

File: java/runtime/src/main/java/org/ray/runtime/task/ArgumentsBuilder.java (hunk 9 above)
Comment (kfstorm) [reply java/runtime/src/main/java/org/ray/runtime/task/ArgumentsBuilder.java]: Done.

File: java/test/src/main/java/org/ray/api/test/ActorTest.java (hunk 10 above)
Comment (kfstorm) [reply java/test/src/main/java/org/ray/api/test/ActorTest.java]: Done.

File: java/runtime/src/main/java/org/ray/runtime/task/ArgumentsBuilder.java (hunk 9 above)
Comment (zhijunfu) [reply java/runtime/src/main/java/org/ray/runtime/task/ArgumentsBuilder.java]: maybe this can be moved to C++ later to ensure it's always consistent

File: src/ray/core_worker/core_worker.cc (hunk 4 above)
Comment (zhijunfu) [reply src/ray/core_worker/core_worker.cc]: This is actually required on my linux dev machine in order to compile, but seems not necessary on ci machines, maybe because of different versions of compilers .etc.

File: src/ray/common/task/task_spec.h (hunk 6 above)
Comment (stephanie-wang) [reply src/ray/common/task/task_spec.h]: I see, thanks! Can you use absl mutex and add thread annotations to the protected fields?

File: src/ray/core_worker/core_worker.h [hunk 12]
```diff
@@ -558,8 +557,11 @@ class CoreWorker {
   // Interface to submit non-actor tasks directly to leased workers.
   std::unique_ptr<CoreWorkerDirectTaskSubmitter> direct_task_submitter_;
 
+  mutable absl::Mutex actor_handles_mutex_;
```
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.h]: Please add a comment explaining why this is needed and what race conditions it protects against.

File: src/ray/common/task/task_spec.h [hunk 13]
```diff
@@ -180,10 +181,13 @@ class TaskSpecification : public MessageWrapper<rpc::TaskSpec> {
   /// Cached scheduling class of this task.
   SchedulingClass sched_cls_id_;
 
+  static absl::Mutex mutex_;
```
Comment (stephanie-wang) [inline src/ray/common/task/task_spec.h]: Please add a comment explaining why this is needed and what race conditions it protects against.

**Review by stephanie-wang — APPROVED** [review]
Thanks!

File: java/runtime/src/main/java/org/ray/runtime/actor/NativeRayActor.java (hunk 11 above)
Comment (stephanie-wang) [reply java/runtime/src/main/java/org/ray/runtime/actor/NativeRayActor.java]: Hmm yeah that sounds right. Can you open an issue so we can track it? I don't think any of us have time to get to it right now.

Comment (raulchen) [reply java/runtime/src/main/java/org/ray/runtime/actor/NativeRayActor.java]: https://github.com/ray-project/ray/issues/6370

**Review by raulchen — APPROVED** [review]

File: src/ray/core_worker/core_worker.h (hunk 12 above)
Comment (kfstorm) [reply src/ray/core_worker/core_worker.h]: Done.

File: src/ray/common/task/task_spec.h (hunk 13 above)
Comment (kfstorm) [reply src/ray/common/task/task_spec.h]: Done.
